pages/blog/graphql-authz.mdx

```diff
@@ -0,0 +1,174 @@
+
+```graphql
+# using @authz directive
+
+type User {
+  id: ID!
+  email: String! @authz(rules: [IsAdmin])
+  posts: [Post!]!
+}
+
+type Post @authz(rules: [CanReadPost]) {
+  id: ID!
+  title: String!
+  body: String!
+  status: Status!
+  author: User!
+}
+
+type Query {
+  users: [User!]! @authz(rules: [IsAuthenticated])
+  post(id: ID!): Post
+}
+
+type Mutation {
+  publishPost(postId: ID!): Post! @authz(rules: [CanPublishPost])
+}
+```
+
+Directives could be used in monolith architecture as well.
+
+If directives are not applicable for some reason, plain JSON auth schema could be used to describe which rules should be executed for certain Object, Interface, Field, Query, or Mutation
+
+```typescript
+// defining auth schema
+
+const authSchema = {
+  Post: { __authz: { rules: ['CanReadPost'] } },
+  User: {
+    email: { __authz: { rules: ['IsAdmin'] } },
+  },
+  Mutation: {
+    publishPost: { __authz: { rules: ['CanPublishPost'] } },
+  },
+  Query: {
+    users: { __authz: { rules: ['IsAuthenticated'] } },
+  },
+  // wildcards are supported. read more: https://github.com/AstrumU/graphql-authz#wildcard-rules
+  '*': {
+    __authz: { rules: ['Reject'] },
+    '*': {
+      __authz: { rules: ['IsAuthenticated'] },
+    },
+  },
+};
+```
+
+## Comparison with GraphQL Shield
+
+[GraphQL Shield](https://github.com/maticzav/graphql-shield) is a great tool for creating authorization layers. Actually, GraphQL AuthZ was inspired by GraphQL Shield! However, it incorporates a different approach compared to GraphQL AuthZ in terms of hooking up authorization logic to the execution process.
+
+The main difference is GraphQL Shield uses field middlewares to execute authorization logic while GraphQL AuthZ wraps the entire execution process. The benefits of the wrapping approach are described in previous paragraphs however there are some drawbacks. For example, with GraphQL AuthZ post-execution rules there is no ability to fail the request earlier because post-execution rules are executed after all resolvers are executed. GraphQL Shield, on the other hand, can fail the request during the execution phase which is earlier than the post-execution phase but later than the pre-execution phase. However post-execution rules have the benefit of accessing results of all field resolvers by design. It means that the post-execution rule attached to a certain field has an access to sibling fields even if these fields are resolved by field resolvers. Also, the post-execution rule attached to a certain field which is resolved by the field resolver has an access to the field value. The middleware approach is missing these possibilities because authorization logic is executed in the context of field resolvers and has access only to the parent object which is the result of the parent field resolver or query. Results of field resolvers of the current and sibling fields are not available to authorization logic at this moment.
```

Comment:
```suggestion
The main difference is GraphQL Shield uses field middleware by wrapping all the `resolver` functions within your GraphQL schema for executing authorization logic during the data-resolving phase, while GraphQL AuthZ wraps the entire `execute` logic.

The benefits of the wrapping approach are described in previous paragraphs, however, there are also some drawbacks. For example, with GraphQL AuthZ post-execution rules there is no ability to fail the request early because post-execution rules are executed after all resolvers are executed. GraphQL Shield, on the other hand, can fail the request during the execution phase which happens before the post-execution phase but later than the pre-execution phase. On the other hand, post-execution rules have the benefit of accessing the resolved value of the field they are specified on and, furthermore, even the sibling fields that are specified via the rules selection set.

The middleware approach of GraphQL shield is missing these possibilities because authorization logic is executed in the context of field resolvers and has access only to the parent object which is the value returned from the parent field resolver. At that stage, the wrapped field resolvers have not been executed yet.
```